README.md
 ![CI](https://github.com/byzer-org/byzer-lang/actions/workflows/build_and_test.yml/badge.svg)   [![License](https://img.shields.io/badge/License-Apache_2.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)

<p align="center">
    <img src="/images/Byzer_Logo.png" alt="drawing"  width="200"/>
</p>

## TOC   
   * [Byzer-Lang](#byzer-lang)
      * [Byzer Code Example](#byzer-code-example)
      * [Byzer Architecture](#byzer-aritchitechture)
   * [Official WebSite](#official-website)
   * [VSCode Extension（MacOS、Linux）](#vscode-extensionmacoslinux)
   * [Docker Sandbox](#docker-sandbox)
      * [Pulling Sandbox Docker Image](#pulling-sandbox-docker-image)
      * [Start Container](#start-container)
   * [Download Byzer](#download-byzer)
   * [Building a Distribution](#building-a-distribution)
      * [Prerequisites](#prerequisites)
      * [Downloading Source Code](#downloading-source-code)
      * [Building Spark 2.4.3 Bundle](#building-spark-243-bundle)
      * [Building Spark 3.1.1 Bundle](#building-spark-311-bundle)
      * [Building without Chinese Analyzer](#building-without-chinese-analyzer)
      * [Building with Aliyun OSS Support](#building-with-aliyun-oss-support)
   * [Deploying](#deploying)
   * [How to contribute to Byzer-Lang](#how-to-contribute-to-byzer-lang)
   * [Contributors](#contributors)
   * [WeChat Group](#wechat-group)

## Byzer-Lang

**Byzer** (former MLSQL) is a low-code, open-sourced and distributed programming language for data pipeline, analytics and AI in cloud native way.

**Deisgn protocol: Everything is a table.** Byzer is a SQL-like language, to simplify data pipeline, analytics and AI, combined with built-in algorithms and extensions.

We believe that everything is a table, a simple and powerful SQL-like language can significantly reduce human efforts of data development without switching different tools.

### Byzer Architecture

![Byzer-lang Arch](images/Byzer-arch.png)

### Byzer Code Example


```sql
load hive.`raw.stripe_discounts` as discounts;
load hive.`raw.stripe_invoice_items` as invoice_items;

select
        invoice_items.*,
        case
            when discounts.discount_type = 'percent'
                then amount * (1.0 - discounts.discount_value::float / 100)
            else amount - discounts.discount_value
        end as discounted_amount

    from invoice_items

    left outer join discounts
        on invoice_items.customer_id = discounts.customer_id
        and invoice_items.invoice_date > discounts.discount_start
        and (invoice_items.invoice_date < discounts.discount_end
             or discounts.discount_end is null)
as joined;



select

        id,
        invoice_id,
        customer_id,
        coalesce(discounted_amount, amount) as discounted_amount,
        currency,
        description,
        created_at,
        deleted_at

    from joined
as final;



set allColumns = "all,wow";

!if ''' split(:allColumns,",")[0] == "all" ''';
   select * from final as final2;
!else;
   select id,invoice from final as final2;
!fi;

select * from final2 as output;
```


## Official WebSite

[https://www.byzer.org](https://www.byzer.org)

## Notebook Support

[byzer-notebook](https://github.com/byzer-org/byzer-notebook)


## VSCode Extension（MacOS、Linux、Windows）

[VSCode IDE Extension](https://github.com/byzer-org/byzer-desktop)

[More document about byzer-lang vscode extension（Chinese version）](https://docs.byzer.org/#/byzer-lang/zh-cn/installation/desktop-installation)

## Docker Sandbox  (With Notebook)

```
export MYSQL_PASSWORD=${1:-root}
export SPARK_VERSION=${SPARK_VERSION:-3.1.1}
export MLSQL_VERSION=${MLSQL_VERSION:-2.2.0-SNAPSHOT}

docker run -d \
-p 3306:3306 \
-p 9002:9002 \
-p 9003:9003 \
-e MYSQL_ROOT_HOST=% \
-e MYSQL_ROOT_PASSWORD="${MYSQL_PASSWORD}" \
--name mlsql-sandbox-${SPARK_VERSION}-${MLSQL_VERSION} \
mlsql-sandbox:${SPARK_VERSION}-${MLSQL_VERSION}
```

Then you can visit `http://127.0.0.1:9002` .


## Download Byzer

* The latest stable version is release-2.2.0
* You can download from [Byzer Download Website](https://download.byzer.org)
* Spark 2.4.3/3.1.1 have been tested

***Naming Convention***

mlsql-engine_${spark_major_version}-${mlsql_version}.tgz
```shell
## Pre-built for Spark 2.4.3
byzer-lang-2.4.3-2.1.0.tar.gz

## Pre-built for Spark 3.1.1
byzer-lang-3.1.1-2.1.0.tar.gz
```  

## Building a Distribution
### Prerequisites
- JDK 8+
- Maven
- Linux or MacOS

### Downloading Source Code
```shell
## Clone the code base
git clone https://github.com/byzer-org/byzer-lang.git
cd byzer-lang
```

### Building Spark 2.4.3 Bundle
```shell
export MLSQL_SPARK_VERSION=2.4
./dev/make-distribution.sh
```

### Building Spark 3.1.1 Bundle
```shell
export MLSQL_SPARK_VERSION=3.0
./dev/make-distribution.sh
```
### Building without Chinese Analyzer
```shell
## Chinese analyzer is enabled by default.
export ENABLE_CHINESE_ANALYZER=false
./dev/make-distribution.sh <spark_version>
```

## Deploying
1. [Download](#Download) or [build a distribution](#Build) 
2. Install Spark and set environment variable SPARK_HOME, make sure Spark version matches that of MLSQL
3. Deploy tgz
- Set environment variable MLSQL_HOME
- Copy distribution tar ball over and untar it

4.Start Byzer in local mode
```shell
cd $MLSQL_HOME
## Run process in background
nohup ./bin/start-local.sh 2>&1 > ./local_mlsql.log &
```
5. Open a browser and type in http://localhost:9003, have fun.

Directory structure
```shell
|-- mlsql
    |-- bin        
    |-- conf       
    |-- data       
    |-- examples   
    |-- libs       
    |-- main       
    |-- README.md  
    |-- LICENSE
    |-- RELEASE
```

## How to contribute to Byzer-Lang

If you are planning to contribute to this repository, please create an issue at [our Issue page](https://github.com/byzer-org/byzer-lang/issues)
even if the topic is not related to source code itself (e.g., documentation, new idea and proposal).

This is an active open source project for everyone,
and we are always open to people who want to use this system or contribute to it.

For more details about how to contribute to the Byzer Org, please refer to [How to Contribute](https://docs.byzer.org/#/byzer-lang/zh-cn/appendix/contribute)


## Contributors

<a href="https://github.com/byzer-org/byzer-lang/graphs/contributors">
  <img src="https://contrib.rocks/image?repo=byzer-org/byzer-lang" />
</a>

Made with [contrib.rocks](https://contrib.rocks).

##  WeChat Group

扫码添加K小助微信号，添加成功后，发送  mlsql  这5个英文字母进群。

![](https://github.com/allwefantasy/mlsql/blob/master/images/dc0f4493-570f-4660-ab41-0e487b17a517.png)



RELEASES.md
## Byzer lang Release Note

Please go to [https://github.com/byzer-org/byzer-lang/releases](https://github.com/byzer-org/byzer-lang/releases) to get the release note.

dev/package.sh
#!/usr/bin/env bash

#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

set -e
set -o pipefail

BASE_DIR=$(cd -P -- "$(dirname -- "${BASH_SOURCE[0]}")/.." && pwd -P)

function exit_with_usage {
  cat << EOF
usage: package
run package command based on different spark version.
Inputs are specified with the following environment variables:

MLSQL_SPARK_VERSION - the spark version, 2.4/3.0 default 2.4
DRY_RUN true|false               default false
DISTRIBUTION true|false          default false
DATASOURCE_INCLUDED true|false   default false
EOF
  exit 1
}

if [[ $@ == *"help"* ]]; then
  exit_with_usage
fi

cd "$BASE_DIR"

MLSQL_SPARK_VERSION=${MLSQL_SPARK_VERSION:-3.0}
DRY_RUN=${DRY_RUN:-false}
DISTRIBUTION=${DISTRIBUTION:-false}
OSS_ENABLE=${OSS_ENABLE:-false}
DATASOURCE_INCLUDED=${DATASOURCE_INCLUDED:-false}
COMMAND=${COMMAND:-package}

ENABLE_JYTHON=${ENABLE_JYTHON=-true}
ENABLE_CHINESE_ANALYZER=${ENABLE_CHINESE_ANALYZER=-false}
ENABLE_HIVE_THRIFT_SERVER=${ENABLE_HIVE_THRIFT_SERVER=-true}

for env in MLSQL_SPARK_VERSION DRY_RUN DISTRIBUTION; do
  if [[ -z "${!env}" ]]; then
    echo "===$env must be set to run this script==="
    echo "===Please run ./dev/package.sh help to get how to use.==="
    exit 1
  fi
done

if [[ ${ENABLE_CHINESE_ANALYZER} = true && ! -f $BASE_DIR/dev/ansj_seg-5.1.6.jar  && ! -f $BASE_DIR/dev/nlp-lang-1.7.8.jar ]]
then
  echo "When ENABLE_CHINESE_ANALYZER=true, ansj_seg-5.1.6.jar && nlp-lang-1.7.8.jar should be in ./dev/"
  exit 1
fi

# before we compile and package, correct the version in MLSQLVersion
#---------------------

unameOut="$(uname -s)"
case "${unameOut}" in
    Linux*)     machine=Linux;;
    Darwin*)    machine=Mac;;
    CYGWIN*)    machine=Cygwin;;
    MINGW*)     machine=MinGw;;
    *)          machine="UNKNOWN:${unameOut}"
esac
echo ${machine}

current_version=$(cat pom.xml|grep -e '<version>.*</version>' | head -n 1 | tail -n 1 | cut -d'>' -f2 | cut -d '<' -f1)

echo "Parse the version of byzer lang, the version is:""$current_version"

MLSQL_VERSION_FILE="./streamingpro-core/src/main/java/tech/mlsql/core/version/MLSQLVersion.scala"

if [[ "${machine}" == "Linux" ]]
then
    sed -i "s/MLSQL_VERSION_PLACEHOLDER/${current_version}/" ${MLSQL_VERSION_FILE}
elif [[ "${machine}" == "Mac" ]]
then
    sed -i '' "s/MLSQL_VERSION_PLACEHOLDER/${current_version}/" ${MLSQL_VERSION_FILE}
else
 echo "Windows  is not supported yet"
 exit 0
fi
#---------------------

BASE_PROFILES="-Ponline  "

if [[ "${ENABLE_HIVE_THRIFT_SERVER}" == "true" ]]; then
  BASE_PROFILES="$BASE_PROFILES -Phive-thrift-server"
fi

if [[ "${ENABLE_JYTHON}" == "true" ]]; then
  BASE_PROFILES="$BASE_PROFILES -Pjython-support"
fi

if [[ "${ENABLE_CHINESE_ANALYZER}" == "true" ]]; then
  BASE_PROFILES="$BASE_PROFILES -Pchinese-analyzer-support"
fi


if [[ "$MLSQL_SPARK_VERSION" == "2.4" ]]; then
  SCALA_BINARY_VERSION="2.11"
  SPARK_VERSION="2.4.3"
else
  SCALA_BINARY_VERSION="2.12"
  SPARK_VERSION="3.1.1"
fi
BASE_PROFILES="$BASE_PROFILES -Pscala-$SCALA_BINARY_VERSION"

BASE_PROFILES="$BASE_PROFILES -Pspark-$MLSQL_SPARK_VERSION.0 -Pstreamingpro-spark-$MLSQL_SPARK_VERSION.0-adaptor"

if [[ ${DISTRIBUTION} == "true" ]];then
BASE_PROFILES="$BASE_PROFILES -Passembly"
else
BASE_PROFILES="$BASE_PROFILES -pl streamingpro-mlsql -am"
fi

export MAVEN_OPTS="-Xmx6000m"

SKIPTEST=""
TESTPROFILE=""

if [[ "${COMMAND}" == "package" ]];then
  BASE_PROFILES="$BASE_PROFILES -Pshade"
fi

if [[ "${COMMAND}" == "package" || "${COMMAND}" == "install" || "${COMMAND}" == "deploy" ]];then
   SKIPTEST="-DskipTests"
fi


if [[ "${COMMAND}" == "test" ]];then
   TESTPROFILE="-Punit-test"
fi

if [[ "${COMMAND}" == "deploy" ]];then
   BASE_PROFILES="$BASE_PROFILES -Prelease-sign-artifacts"
   BASE_PROFILES="$BASE_PROFILES -Pdisable-java8-doclint"
fi

if [[ "${OSS_ENABLE}" == "true" ]];then
   BASE_PROFILES="$BASE_PROFILES -Paliyun-oss"
fi

if [[ "$DATASOURCE_INCLUDED" == "true" ]];then
   BASE_PROFILES="$BASE_PROFILES -Punit-test"
fi

if [[ ${DRY_RUN} == "true" ]];then

cat << EOF
mvn clean ${COMMAND}  ${SKIPTEST} ${BASE_PROFILES}  ${TESTPROFILE}
EOF

else
cat << EOF
mvn clean ${COMMAND}  ${SKIPTEST} ${BASE_PROFILES}  ${TESTPROFILE}
EOF
mvn clean ${COMMAND}  ${SKIPTEST} ${BASE_PROFILES} ${TESTPROFILE}
fi
